src/Effects/EffectMgr.cpp

```diff
@@ -117,6 +120,98 @@ static const rtosc::Ports local_ports = {
                 }
             }
         }},
+    {"numerator::i", rShort("num") rDefault(0) rLinear(0,99)
+        rProp(parameter) rDoc("Numerator of ratio to bpm"), NULL,
+        [](const char *msg, rtosc::RtData &d)
+        {
+            EffectMgr *eff = (EffectMgr*)d.obj;
+            if(rtosc_narguments(msg)) {
+                int val = rtosc_argument(msg, 0).i;
+                if (val) {
+                    eff->numerator = val;
+                    int Pdelay, Pfreq;
+                    float freq;
+                    if(eff->denominator) {
+                        switch(eff->nefx) {
+                        case 2: // Echo
+                            // invert:
+                            // delay = (Pdelay / 127.0f * 1.5f); //0 .. 1.5 sec
+                            Pdelay = (int)roundf((20320.0f / (float)eff->time->tempo) * 
+                                                 ((float)eff->numerator / (float)eff->denominator));
+                            if (eff->numerator&&eff->denominator)
+                                eff->seteffectparrt(2, Pdelay);
+                            break;
+                        case 3: // Chorus
+                        case 4: // Phaser
+                        case 5: // Alienwah
+                        case 8: // DynamicFilter
+                            freq =  ((float)eff->time->tempo / 15.0f * 
+                                     (float)eff->numerator / (float)eff->denominator);
+                            // invert:
+                            // (powf(2.0f, Pfreq / 127.0f * 10.0f) - 1.0f) * 0.03f
+                            Pfreq = (int)roundf(logf((freq/0.03f)+1.0f)/LOG_2 * 12.7f);
+                            if (eff->numerator&&eff->denominator)
+                                eff->seteffectparrt(2, Pfreq);
+                            break;
+                        case 1: // Reverb
+                        case 6: // Distorsion
+                        case 7: // EQ
+                        default:
+                            break;
+                        }
+                    }
+                }
+                d.broadcast(d.loc, "i", val);
+            } else {
+                d.reply(d.loc, "i", eff->numerator); 
+            }
+        }},
+    {"denominator::i", rShort("dem") rDefault(4) rLinear(1,99)
+        rProp(parameter) rDoc("Denominator of ratio to bpm"), NULL,
+        [](const char *msg, rtosc::RtData &d)
+        {
+            EffectMgr *eff = (EffectMgr*)d.obj;
+            if(rtosc_narguments(msg)) {
+                int val = rtosc_argument(msg, 0).i;
+                if (val) {
```

Comment:
I'd change this to if(val > 0) to prevent negative values (note the GUI will not send negative values, but the users can with the existing receiver code)

Reply:
ok